service/history/api/respondworkflowtaskcompleted/workflow_task_completed_handler.go

```diff
@@ -478,7 +478,9 @@ func (handler *workflowTaskCompletedHandler) handleCommandScheduleActivity(
 
 	// TODO: versioning 3 allows eager activity dispatch for both pinned and unpinned workflows, no
 	// special consideration is need. Remove the versioning logic from here. [cleanup-old-wv]
-	oldVersioningUsed := handler.mutableState.GetMostRecentWorkerVersionStamp().GetUseVersioning()
+	oldVersioningUsed := handler.mutableState.GetMostRecentWorkerVersionStamp().GetUseVersioning() &&
+		// for V3 versioning it's ok to dispatch eager activities
+		handler.mutableState.GetEffectiveVersioningBehavior() == enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED
 	newVersioningUsed := handler.mutableState.GetExecutionInfo().GetAssignedBuildId() != ""
```

Comment:
is `newVersioningUsed` correct? it seems like this only checks that the WF never used versioning v2.

Reply:
I think it is correct. once a wf uses v2 it cannot go back! in here I just want to exclude v3 from the check. not touching v1-2 suff.